src/Microsoft.VisualStudio.SDK.Analyzers.Tests/VSSDK006CheckServicesExistAnalyzerTests.cs

```diff
@@ -0,0 +1,636 @@
+        Assumes.Present(this.svc);
+    }
+
+    void Foo() {
+        this.svc.BeginDesignTimeBuild();
+    }
+}
+";
+
+        this.VerifyCSharpDiagnostic(test, this.CreateDiagnostic(9, 9, 8, (13, 9, 8)));
+        this.VerifyCSharpFix(test, fix);
+    }
+
+    [Fact]
+    public void GetService_DirectlyUsed()
+    {
+        var test = @"
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+class Test : Package {
+    protected override void Initialize() {
+        base.Initialize();
+        ((IVsBuildManagerAccessor)this.GetService(typeof(SVsBuildManagerAccessor))).BeginDesignTimeBuild();
+    }
+}
+";
+
+        this.VerifyCSharpDiagnostic(test, this.CreateDiagnostic(8, 35, 15));
+        this.VerifyNoCSharpFixOffered(test);
+    }
+
+    [Fact]
+    public void GetService_DirectlyUsed_WithConditionalMemberAccess()
+    {
+        var test = @"
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+class Test : Package {
+    protected override void Initialize() {
+        base.Initialize();
+        ((IVsBuildManagerAccessor)this.GetService(typeof(SVsBuildManagerAccessor)))?.BeginDesignTimeBuild();
+    }
+}
+";
+
+        this.VerifyCSharpDiagnostic(test);
+    }
+
+    [Fact]
+    public void GetServiceAsync_DirectlyUsed()
+    {
+        var test = @"
+using System;
+using System.Threading;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Task = System.Threading.Tasks.Task;
+
+class Test : AsyncPackage {
+    protected override async Task InitializeAsync(CancellationToken cancellationToken, IProgress<ServiceProgressData> progress) {
+        await base.InitializeAsync(cancellationToken, progress);
+        ((IVsBuildManagerAccessor)await this.GetServiceAsync(typeof(SVsBuildManagerAccessor))).BeginDesignTimeBuild();
+    }
+}
+";
+
+        this.VerifyCSharpDiagnostic(test, this.CreateDiagnostic(11, 41, 20));
+        this.VerifyNoCSharpFixOffered(test);
+    }
+
+    [Fact]
+    public void LocalAssigned_CheckedByThrow_GetService_ThenUsed()
+    {
+        var test = @"
+using Microsoft;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+class Test : Package {
+    protected override void Initialize() {
+        base.Initialize();
+        var svc = this.GetService(typeof(SVsBuildManagerAccessor)) as IVsBuildManagerAccessor;
+        Assumes.Present(svc);
+        svc.BeginDesignTimeBuild();
+    }
+}
+";
+        this.VerifyCSharpDiagnostic(test);
+    }
+
+    [Fact]
+    public void LocalAssigned_CheckedByIf_GetService_ThenUsed()
+    {
+        var test = @"
+using Microsoft;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+class Test : Package {
+    protected override void Initialize() {
+        base.Initialize();
+        var svc = this.GetService(typeof(SVsBuildManagerAccessor)) as IVsBuildManagerAccessor;
+        if (svc != null) {
```

Comment:
what happen if user has svc?.BeginDes….();

will it still ask to add Assumes.Present?

Reply:
I added a test for this, and it Just Worked. :)